### PR TITLE
Zero alloc annotation: assume a function never returns normally

### PR DESCRIPTION
**Pull request by gretay-js**
Support for `[@zero_alloc assume never_returns_normally]` function annotation. The intended use is for situations in which the analysis cannot figure out statically that a function does not return normally, for example:

```
let[@inline never][@zero_alloc assume never_returns_normally] failwithf fmt =
  Printf.ksprintf (fun s () -> failwith s) fmt

let[@inline never][@zero_alloc assume never_returns_normally] invalid_argf fmt =
  Printf.ksprintf (fun s () -> invalid_arg s) fmt

let[@zero_alloc] foo x = failwithf "%d" x
let[@zero_alloc] bar x y = invalid_argf "%d" (x+y)
```

The first commit refactors `Lambda.check_attribute`: instead of `Check { assume=true; ...}` use a separate constructor for `Assume` to make it easier to add a new field to only one of `Check` or `Assume`.  Then, we add a the field `never_returns_normally` to `Assume`. (A subsequent PR will also add an `opt` field to `Check`. ) 
The new field is used in `checkmach` to set `nor` field to `Bot` instead of the current value of `Safe` that is used to express `assume` annotation on a function.  

The above example currently doesn't work without `[@inline never]` because `never_returns_normally` annotation is not propagated as part of  `Debuginfo.t` in the same was `assume_zero_alloc`. I am planning to make a separate PR to  fix it but in the meantime the current PR is sufficient to handle the [common use case from Base](https://github.com/janestreet/base/blob/fbb560c9fe1e3d82c7ecca18a4b56eac9ad86886/src/printf.ml#L6-L7).

### CODE REVIEW

**Comment by mshinwell** [comment]
@xclerc if you could review the whole of this it would be appreciated

**Comment by gretay-js** [comment]
@xclerc  I'll be happy to split refactoring commits into a separate feature, if it's easier to review.

File: middle_end/flambda2/terms/check_attribute.ml [hunk 1]
```diff
@@ -35,27 +40,42 @@ let print ppf t =
...
-  | (Default_check | Ignore_assert_all _ | Check _), _ -> false
+  | ( Check { property = p1; strict = s1; loc = loc1 },
+      Check { property = p2; strict = s2; loc = loc2 } ) ->
+    Property.equal p1 p2 && Bool.equal s1 s2 && loc1 = loc2
+  | ( Assume
+        { property = p1; strict = s1; never_returns_normally = n1; loc = loc1 },
+      Assume
+        { property = p2; strict = s2; never_returns_normally = n2; loc = loc2 }
+    ) ->
+    Property.equal p1 p2 && Bool.equal s1 s2 && Bool.equal n1 n2 && loc1 = loc2
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/terms/check_attribute.ml:76]: Yes we should - let's fix that here please.

File: ocaml/lambda/translattribute.ml [hunk 2]
```diff
@@ -251,11 +251,17 @@ let parse_property_attribute attr property =
...
-          ["strict"], Check { property; strict = true; assume = false; loc; };
-          ["assume"; "strict"], Check { property; strict = true; assume = true; loc; };
+          ["assume"],
+          Assume { property; strict = false; never_returns_normally = false; loc; };
+          ["strict"], Check { property; strict = true; loc; };
+          ["assume"; "strict"],
+          Assume { property; strict = true; never_returns_normally = false; loc; };
+          ["assume"; "never_returns_normally"],
+          Assume { property; strict = false; never_returns_normally = true; loc; };
+          ["assume"; "strict"; "never_returns_normally"],
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply ocaml/lambda/translattribute.ml:263]: This is covered by sorting the list in `parse_ids_payload`.

**Comment by xclerc** [comment]
(Orthogonally, even if we promote never-returns
to a property, there is certainly an argument to
do so in another pull request.)

File: middle_end/flambda2/terms/check_attribute.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (gretay-js) [reply middle_end/flambda2/terms/check_attribute.ml]: Fixed.

**Comment by gretay-js** [comment]
> (Orthogonally, even if we promote never-returns
to a property, there is certainly an argument to
do so in another pull request.)

Thank you. I'll make a follow-up PR to try this out.

**Review by mshinwell — APPROVED** [review]

**Pull request merged**